src/deepsparse/transformers/pipelines/text_generation.py

```diff
@@ -656,25 +675,33 @@ def engine_forward(
                         finished_reason.append(FinishReason.STOP)
                         break
 
-                    # TODO: Add any generic callback reason?
                     if callback is not None and callback(token) is False:
                         _LOGGER.debug(
                             "callback %s returned False, stopping generation."
                             % callback.__qualname__
                         )
+                        finished_reason.append(FinishReason.CALLBACK)
                         break
 
                     if len(generated_tokens) == max_tokens:
                         finished_reason.append(FinishReason.LENGTH)
 
-            if streamer is not None:
-                streamer.end()
+                    if self.streaming:
+                        yield (numpy.array([token]), numpy.array([logits]), [None])
 
-        return (
-            numpy.array([generated_tokens]),
-            numpy.concatenate(generated_logits, axis=1),
-            finished_reason,
-        )
+                if self.streaming:
+                    yield (
+                        numpy.array([token]),
+                        numpy.array([logits]),
+                        [finished_reason[-1]],
+                    )
+
+        if not self.streaming:
```

Comment:
why do we need to yield instead of return?

Reply:
any function with a yield will always return a generator but refactoring out the streaming code to be a separate function seems to a workaround